pkg/scheduler/queue/queue.go

```diff
@@ -8,75 +8,90 @@ Licensed under the MIT license.
 package queue
 
 import (
+	"time"
+
 	"k8s.io/client-go/util/workqueue"
 )
 
-// ClusterPolicySnapshotKey is the unique identifier (its name) for a ClusterPolicySnapshot stored in a scheduling queue.
-type ClusterPolicySnapshotKey string
-
-// ClusterPolicySnapshotKeySchedulingQueueWriter is an interface which allows sources, such as controllers, to add
-// ClusterPolicySnapshots to the scheduling queue.
-type ClusterPolicySnapshotKeySchedulingQueueWriter interface {
-	Add(cpsKey ClusterPolicySnapshotKey)
+// ClusterResourcePlacementKey is the unique identifier (its name) for a ClusterResourcePlacement checked
+// into a scheduling queue.
+type ClusterResourcePlacementKey string
+
+// ClusterResourcePlacementSchedulingQueueWriter is an interface which allows sources, such as controllers, to add
+// ClusterResourcePlacementKeys to the scheduling queue.
+type ClusterResourcePlacementSchedulingQueueWriter interface {
+	// Add adds a ClusterResourcePlacementKey to the work queue.
+	//
+	// Note that this bypasses the rate limiter.
+	Add(crpKey ClusterResourcePlacementKey)
+	// AddRateLimited adds a ClusterResourcePlacementKey to the work queue after the rate limiter (if any)
+	// says that it is OK.
+	AddRateLimited(crpKey ClusterResourcePlacementKey)
+	// AddAfter adds a ClusterResourcePlacementKey to the work queue after a set duration.
+	AddAfter(crpKey ClusterResourcePlacementKey, duration time.Duration)
 }
 
-// ClusterPolicySnapshotSchedulingQueue is an interface which queues ClusterPolicySnapshots for the scheduler to schedule.
-type ClusterPolicySnapshotKeySchedulingQueue interface {
-	ClusterPolicySnapshotKeySchedulingQueueWriter
+// ClusterResourcePlacementSchedulingQueue is an interface which queues ClusterResourcePlacements for the scheduler
+// to consume; specifically, the scheduler finds the latest scheduling policy snapshot associated with the
+// ClusterResourcePlacement.
+type ClusterResourcePlacementSchedulingQueue interface {
+	ClusterResourcePlacementSchedulingQueueWriter
 
 	// Run starts the scheduling queue.
 	Run()
 	// Close closes the scheduling queue immediately.
 	Close()
 	// CloseWithDrain closes the scheduling queue after all items in the queue are processed.
 	CloseWithDrain()
-	// NextClusterPolicySnapshotKey returns the next-in-line ClusterPolicySnapshot key for the scheduler to schedule.
-	NextClusterPolicySnapshotKey() (key ClusterPolicySnapshotKey, closed bool)
-	// Done marks a ClusterPolicySnapshot key as done.
-	Done(cpsKey ClusterPolicySnapshotKey)
+	// NextClusterResourcePlacementKey returns the next-in-line ClusterResourcePlacementKey for the scheduler to consume.
+	NextClusterResourcePlacementKey() (key ClusterResourcePlacementKey, closed bool)
+	// Done marks a ClusterResourcePlacementKey as done.
+	Done(crpKey ClusterResourcePlacementKey)
+	// Forget untracks a ClusterResourcePlacementKey from rate limiter(s) (if any) set up with the queue.
+	Forget(crpKey ClusterResourcePlacementKey)
```

Comment:
What's the difference between these two?

Reply:
Hi Ryan! This is the same as a regular rate limiter queue, where `Done` marks a key as no longer in processing and `Forget` is to untrack the key from the rate limiter. They cannot be merged into one as there are cases where a key must be retried with approval from the rate limiter.

Reply:
Hi Ryan! I will merge this PR first; if there's any concern over Done() and Forget(), please let me know.